build-system/compile/json-schema/index.js

```diff
@@ -182,27 +182,254 @@ function transformAjvCode(code, scope, config) {
+  function treeShakeSchema(binding) {
+    const {path, referencePaths} = binding;
+    if (
+      !binding.constant ||
+      !path.isVariableDeclarator() ||
+      !t.isIdentifier(path.node.id) ||
+      !schemaVariableDeclaratorRe.test(path.node.id.name)
+    ) {
+      return;
+    }
+    const init = path.get('init');
+    if (!init.isObjectExpression()) {
+      return;
+    }
+    if (!referencePaths.length) {
+      path.remove();
+      return;
+    }
+    // Re-construct schema from references
+    const original = init.evaluate().value;
+    const replacement = {};
+    if (original == null) {
+      // original is dynamic, deopt
+      return;
+    }
+    /**
+     * Maps container objects to their member expressions.
+     * A container object is one that we create during traversal, and is never
+     * referenced directly. At the point a container is referenced, its mapping
+     * is set as `false`.
+     * @type {Map<Object, Set<babel.NodePath<babel.types.MemberExpression>> | false>}
+     */
+    const containers = new Map();
+    const queue = [...referencePaths];
+    for (const referencePath of queue) {
+      // Navigate MemberExpression to find deepest key to insert.
+      // We create necessary depth in replacement object as we go along.
+      let key;
+      let originalAt = original;
+      let replacementAt = replacement;
+      let containersAt;
+      let memberExpression;
+      let {parentPath} = referencePath;
+      while (replacementAt !== originalAt && parentPath?.isMemberExpression()) {
+        memberExpression = parentPath;
+        parentPath = memberExpression.parentPath;
+        const nextKey = evaluatePropertyKey(memberExpression.get('property'));
+        if (nextKey == null) {
+          break;
+        }
+        if (key != null) {
+          if (!replacementAt[key]) {
+            // Even though the original container may be an array, we first
+            // create it as an object in order to insert sparse keys.
+            // Container objects are all converted to arrays later.
+            replacementAt[key] = {};
+          }
+          replacementAt = replacementAt[key];
+          originalAt = originalAt[key];
+        }
+        key = nextKey;
+        containersAt = containers.get(replacementAt);
+        if (containersAt == null) {
+          containersAt = new Set();
+          containers.set(replacementAt, containersAt);
+        }
+        if (containersAt !== false) {
+          containersAt.add(memberExpression);
+        }
+      }
+      // deopt full schema if we can't resolve at least one key deep
+      if (key == null) {
+        return;
+      }
+      // already inserted as-is
+      if (replacementAt[key] === originalAt[key]) {
+        continue;
+      }
+      // Fill indirect references with fully qualified paths, and add those
+      // to the queue.
+      if (
+        memberExpression &&
+        parentPath?.isVariableDeclarator() &&
+        parentPath.node.init &&
```

Comment:
Is this supposed to be checking the member expression is the init?

Reply:
Yea. We probably don't need to check that `init` is present.